src/main/scala/com/microsoft/hyperspace/HyperspaceSparkSessionExtension.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace
+
+import org.apache.spark.sql.SparkSessionExtensions
+
+import com.microsoft.hyperspace.index.execution.BucketUnionStrategy
+import com.microsoft.hyperspace.index.rules.ApplyHyperspace
+
+/**
+ * An extension for Spark SQL to activate Hyperspace.
+ *
+ * Example to run a `spark-shell` with Hyperspace enabled:
+ * {{{
+ *   spark-shell -c spark.sql.extensions=com.microsoft.hyperspace.HyperspaceSparkSessionExtension
+ * }}}
+ *
+ * Example to create a `SparkSession` with Hyperspace enabled:
+ * {{{
+ *    val spark = SparkSession
+ *       .builder()
+ *       .appName("...")
+ *       .master("...")
+ *       .config("spark.sql.extensions", "com.microsoft.hyperspace.HyperspaceSparkSessionExtension")
+ *       .getOrCreate()
+ * }}}
+ */
+class HyperspaceSparkSessionExtension extends (SparkSessionExtensions => Unit) {
```

Comment:
How about simplifying the enable/disable process by always putting the `ApplyHyperspace` rule in the optimizer and just toggling a flag to control whether the rule is applied or not? The overall picture would be like this:

1. Users are required to include `com.microsoft.hyperspace.HyperspaceSparkSessionExtension` in `spark.sql.extensions` to apply Hyperspace rules.
    1. See below for backward compatibility.
    2. `HyperspaceSparkSessionExtension` is responsible for putting `ApplyHyperspace` at the end of `experimentalMethods.extraOptimizations`. This can be done as the current PR, which is a little bit hacky but works. Ideally, there should be a way for extensions to access `extraOptimizations`. We can propose this change to Spark if we want.
2. Users use a config `spark.hyperspace.apply.enabled` to enable/disable the Hyperspace rule application.
    1. The value is `true` by default. Therefore, just setting `spark.sql.extensions` correctly is enough to use Hyperspace.
    2. `ApplyHyperspace` checks this config value and does nothing if it is `false`. 
    3. We might extend this mechanism to allow users to enable/disable individual rules, e.g. disable applying data skipping indexes by setting `spark.hyperspace.apply.dataskipping.enabled` to `false`, or including the rule name in `spark.hyperspace.apply.excludedRules` (for this to be usable, rule names should be user-friendly and comprehensible).
3. For backward compatibility, continue providing `sparkSession.enable/disableHyperspace()`, although they should be deprecated in favor of the new mechanism.
    1. `enableHyperspace()` puts `ApplyHyperspace` at the end of `extraOptimizations` if it's not already there and sets `spark.hyperspace.apply.enabled` to `true`.
    2. `disableHyperspace()` sets `spark.hyperspace.apply.enabled` to `false`. It doesn't remove the rule.

Reply:
@clee704 Thanks for the suggestion. To check whether I understand correctly, in the above suggestion, the existing program users don't have to change their code but they should at least add `com.microsoft.hyperspace.HyperspaceSparkSessionExtension` to their spark session, right?

Reply:
No, they don't have to, until we deprecate and remove `sparkSession.enable/disableHyperspace()`.

Reply:
As for the naming of the configs, how about `spark.hyperspace.optimizer.*`, as `apply` or `applyHyperspace` might be unclear to users?

Reply:
> puts ApplyHyperspace at the end of extraOptimizations if it's not already there

Oh I understand it now. I missed the above part :)

Ok. I will make change as you suggested.